### PR TITLE
support stale read in plan

### PR DESCRIPTION
**Pull request by YANGGMM**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #14784 

## What this PR does / why we need it:
support stale read in plan

### CODE REVIEW

**Comment by matrix-meow** [comment]
@YANGGMM Thanks for your contributions!

**Here are review comments for file pkg/frontend/authenticate.go:**
### Pull Request Review:

#### Title and Body:
The title of the pull request is clear and concise, indicating that the purpose of the changes is to "support stale read in plan." The body of the pull request provides additional context by referencing the related issue #14784. It would be beneficial to include a brief description of what "stale read" means in this context to help reviewers understand the significance of the changes.

#### Changes in `pkg/frontend/authenticate.go`:
1. **Addition of SQL Query for Snapshot Timestamp Retrieval:**
   - A new SQL query `getSnapshotTsWithSnapshotNameFormat` has been added to retrieve the timestamp associated with a snapshot name.
   - A new function `getSqlForGetSnapshotTsWithSnapshotName` has been implemented to generate the SQL query for retrieving snapshot timestamps based on the snapshot name.
   - The function `getSqlForCheckStageStatus` has been modified to include context and error handling for retrieving snapshot timestamps.

2. **New Function for Resolving Snapshot Timestamp:**
   - A new function `doResolveSnapshotTsWithSnapShotName` has been introduced to handle the resolution of snapshot timestamps.
   - This function performs validation on the input snapshot name, retrieves the SQL query for snapshot timestamp retrieval, executes the query, and processes the result to obtain the snapshot timestamp.
   - Error handling is implemented throughout the function to manage potential issues during the process.

#### Suggestions for Improvement:
1. **Input Sanitization and Validation:**
   - Ensure that proper input validation and sanitization are applied to prevent SQL injection vulnerabilities. Consider using parameterized queries or input validation functions to sanitize user inputs.

2. **Error Handling and Logging:**
   - Enhance error handling by providing more descriptive error messages to aid in debugging and troubleshooting.
   - Consider logging relevant information during error conditions to facilitate error analysis and resolution.

3. **Code Optimization:**
   - Consolidate common error handling logic into reusable functions to reduce code duplication and improve maintainability.
   - Review the code to identify opportunities for refactoring or simplification to enhance readability and maintainability.

4. **Security Considerations:**
   - Conduct a thorough security review to identify and address any potential security vulnerabilities, especially related to database queries and input validation.
   - Implement appropriate access controls and permissions to restrict unauthorized access to sensitive data.

5. **Testing and Validation:**
   - Ensure comprehensive testing is conducted to validate the functionality of the new code additions and modifications.
   - Include test cases that cover different scenarios, including valid and invalid inputs, to verify the correctness of the implemented logic.

By addressing the above suggestions and considerations, the pull request can be enhanced in terms of security, robustness, and maintainability. It is essential to prioritize security measures and adhere to best practices when handling sensitive data and executing database queries.

**Here are review comments for file pkg/frontend/authenticate_test.go:**
### Pull Request Review:

#### Title:
The title of the pull request is clear and concise, indicating that the purpose of the PR is to "support stale read in plan."

#### Body:
The body of the pull request provides information about the type of PR, the specific issue it fixes (issue #14784), and the reason for the PR, which is to support stale read in the plan. The body could be improved by providing more context or details about the changes made.

#### Changes in `pkg/frontend/authenticate_test.go`:
1. **Addition of Test Function**:
   - Two new test functions `TestDoResolveSnapshotTsWithSnapShotName` have been added to the file.
   - Each test function sets up a test scenario, executes a function `doResolveSnapshotTsWithSnapShotName`, and asserts the result using `convey.So`.
   - The tests cover scenarios with different result sets for SQL queries related to snapshots.

2. **Code Structure**:
   - The code structure of the test functions seems well-segmented and follows a standard testing pattern using `convey.Convey`.
   - Proper setup and teardown of resources are handled using `defer` statements.

3. **Concerns**:
   - **Code Duplication**: There is significant duplication of code between the two test functions. This duplication can lead to maintenance issues if changes are required in the setup logic.
   - **Magic Strings**: The SQL queries like "begin;", "commit;", and "rollback;" are directly used as keys in `bh.sql2result`. It would be better to define them as constants to avoid potential errors due to typos.
   - **Hardcoded Date**: The date "2024-01-01 00:00:00" is hardcoded in one of the test scenarios. It's advisable to use dynamic or configurable values to improve test reliability.

#### Suggestions for Improvement:
1. **Code Reusability**:
   - Refactor the common setup logic into a separate function that can be reused by both test functions to avoid code duplication.
   
2. **Constants for SQL Queries**:
   - Define constants for SQL queries used as keys in `bh.sql2result` to enhance code readability and maintainability.
   
3. **Dynamic Date Generation**:
   - Instead of hardcoding the date "2024-01-01 00:00:00", consider using a dynamic date generation method to ensure test stability over time.

4. **Enhance Test Coverage**:
   - Ensure that the test scenarios cover edge cases and potential failure scenarios to improve the robustness of the test suite.

By addressing the mentioned points, the code quality, maintainability, and reliability of the test suite can be improved.

**Here are review comments for file pkg/frontend/compiler_context.go:**
### Pull Request Review:

#### Title:
The title of the pull request is clear and concise, indicating that the purpose of the changes is to support stale read in the plan.

#### Body:
The body of the pull request provides relevant information such as the type of PR, the specific issue it addresses, and the reason for the changes. It mentions that the PR is a feature addition to support stale read in the plan and references the related issue #14784.

#### Changes in `compiler_context.go`:
1. **Addition of `ResolveSnapshotTsWithSnapShotName` function:**
   - A new function `ResolveSnapshotTsWithSnapShotName` has been added to the `TxnCompilerContext` struct in `compiler_context.go`.
   - This function calls `doResolveSnapshotTsWithSnapShotName` with parameters obtained from the `TxnCompilerContext`.
   - **Issue:** The implementation of `ResolveSnapshotTsWithSnapShotName` directly calls an external function `doResolveSnapshotTsWithSnapShotName` without providing details about its implementation or origin.
   - **Suggestion:** It is essential to ensure that the `doResolveSnapshotTsWithSnapShotName` function is secure and properly handles input parameters to prevent any security vulnerabilities. Additionally, consider adding comments or documentation to explain the purpose and behavior of this function.

#### Suggestions for Improvement:
1. **Security Concerns:**
   - Validate the input parameters passed to `doResolveSnapshotTsWithSnapShotName` to prevent any potential security vulnerabilities such as injection attacks.
   - Consider adding input validation and sanitization checks to ensure the function behaves as expected and does not introduce security risks.

2. **Documentation:**
   - Provide detailed comments for the `ResolveSnapshotTsWithSnapShotName` function to explain its purpose, expected input parameters, and return values.
   - Include information about the `doResolveSnapshotTsWithSnapShotName` function, its source, and any potential side effects or dependencies.

3. **Optimization:**
   - Ensure that the `ResolveSnapshotTsWithSnapShotName` function follows best practices for performance and efficiency.
   - Consider optimizing the function implementation if necessary to improve code readability and maintainability.

By addressing the security concerns, adding comprehensive documentation, and optimizing the code, the pull request can be enhanced to ensure the reliability and maintainability of the codebase.

**Here are review comments for file pkg/pb/plan/plan.pb.go:**
### Pull Request Review:

#### Title:
The title of the pull request is clear and concise, indicating that the purpose of the changes is to support stale read in the plan.

#### Body:
The body of the pull request provides some context by mentioning the type of PR (Feature), the specific issue it addresses (issue #14784), and the reason for the changes ("support stale read in plan"). It would be helpful to have more details on how stale read is being supported in the plan.

#### Changes in `pkg/pb/plan/plan.pb.go`:
Since there is no actual code changes shown in the diff, it is not possible to provide a detailed review of the modifications made in the file. It is recommended to ensure that the changes are included in the diff for a comprehensive review.

### Suggestions for Improvement:

1. **Provide Detailed Changes:**
   - Ensure that the actual code changes are included in the diff so that reviewers can analyze the modifications made to support stale read in the plan.

2. **Clarify Implementation Details:**
   - Add more information in the body of the pull request about how stale read is being implemented in the plan to give a better understanding to reviewers.

3. **Security Considerations:**
   - When implementing stale read functionality, ensure that proper security measures are in place to prevent any data integrity issues or unauthorized access.

4. **Optimization:**
   - If the changes involve performance implications, consider mentioning any optimizations or trade-offs made to support stale read efficiently.

By addressing the points mentioned above, the pull request can be improved in terms of clarity, completeness, and security considerations.

**Here are review comments for file pkg/sql/compile/sql_executor_context.go:**
### Pull Request Review:

#### Title and Body:
The title of the pull request is clear and concise, indicating that the purpose of the PR is to "support stale read in plan." The body of the PR provides additional context by referencing the related issue #14784. However, it lacks detailed information on the changes being made and the rationale behind them. It would be beneficial to include a brief explanation of what "support stale read in plan" entails and why it is necessary for the project.

#### Changes in `sql_executor_context.go`:
1. **Adding `ResolveSnapshotTsWithSnapShotName` Function:**
   - The addition of the `ResolveSnapshotTsWithSnapShotName` function seems to be incomplete as it currently contains a panic statement. This can lead to unexpected behavior and is not a proper implementation of the function.
   - **Suggestion:** Instead of using `panic`, the function should be implemented to handle the resolution of snapshot timestamps based on the provided snapshot name. Proper error handling and logic should be added to ensure the function works as intended.

2. **Update in `IsPublishing` and `SetQueryingSubscription` Functions:**
   - The existing functions `IsPublishing` and `SetQueryingSubscription` have `panic` statements indicating they are not supported in the internal SQL executor.
   - **Suggestion:** If these functions are not intended to be used in the internal SQL executor, consider either removing them or providing a proper implementation that aligns with the internal executor's requirements. Using `panic` is not a suitable approach for handling unsupported functionality.

#### Security Concern:
1. **Panic Statements:**
   - The liberal use of `panic` statements in the code can lead to unexpected crashes and potential security vulnerabilities. It is crucial to handle errors and unsupported operations gracefully rather than abruptly terminating the program with a panic.
   - **Suggestion:** Replace `panic` statements with proper error handling mechanisms such as returning errors or logging messages to ensure the stability and security of the application.

### Overall Suggestions:
- Provide a more detailed explanation in the PR body regarding the purpose and impact of the changes.
- Implement the `ResolveSnapshotTsWithSnapShotName` function with proper logic instead of using `panic`.
- Review the usage of `panic` statements in the codebase and replace them with appropriate error handling mechanisms.
- Ensure that all functions are properly implemented or removed if they are not intended for use in the internal SQL executor.

By addressing these issues and suggestions, the code quality, maintainability, and security of the project can be significantly improved.

**Here are review comments for file pkg/sql/parsers/dialect/mysql/mysql_sql.go:**
### Pull Request Review:

#### Title:
The title of the pull request is clear and concise, indicating that the purpose of the change is to support stale read in the plan.

#### Body:
The body of the pull request provides some context by mentioning the type of PR (Feature), the related issue (#14784), and the purpose of the change (support stale read in plan). It would be beneficial to include more details about what "stale read" means in this context and how it will be implemented.

#### Changes in `pkg/sql/parsers/dialect/mysql/mysql_sql.go`:
Since there is no actual code changes shown in the diff, it is challenging to provide specific feedback on the implementation details. However, based on the description provided, here are some general suggestions and potential issues to consider:

1. **Lack of Code Changes:**
   - It appears that the diff does not show any actual code changes in the `mysql_sql.go` file. It is essential to ensure that the changes related to supporting stale reads are correctly included in the diff.

2. **Implementation of Stale Read:**
   - Ensure that the implementation of stale read functionality follows best practices and does not introduce any performance or security vulnerabilities.
   - Consider how stale read will be handled in the SQL parser and how it will affect query execution.

3. **Testing:**
   - It is crucial to include comprehensive tests to cover the new functionality related to stale reads. This ensures that the feature works as expected and helps prevent regressions.

4. **Documentation:**
   - Update relevant documentation to reflect the changes related to supporting stale reads. This helps other developers understand how to use this new feature.

5. **Code Quality:**
   - If there are no code changes in the diff, consider if there are any missed modifications or if the changes are located in a different file.

6. **Optimization:**
   - If the implementation involves significant changes or performance considerations, ensure that the code is optimized and does not introduce unnecessary overhead.

### Overall, it is important to ensure that the implementation of supporting stale read in the plan is thorough, well-tested, and documented. If there are no code changes in the provided diff, double-check to confirm that the changes are correctly included and address any potential issues that may arise during the implementation.

**Here are review comments for file pkg/sql/parsers/dialect/mysql/mysql_sql.y:**
### Pull Request Review:

#### Title and Body:
The title of the pull request is clear and concise, indicating that the purpose is to "support stale read in plan." The body provides additional context by mentioning the related issue number (#14784) and the need for supporting stale read in the plan. It would be beneficial to include a bit more detail in the body about what "stale read" means and how it impacts the functionality.

#### Changes in `pkg/sql/parsers/dialect/mysql/mysql_sql.y`:
1. **Addition of Stale Read Support:**
   - The changes introduce support for stale read in the parser by adding new rules for handling timestamps and snapshots in table names.
   - The modifications include defining new types (`ATTIMESTAMPTIME` and `ATTIMESTAMPSNAPSHOT`) to differentiate between timestamp and snapshot reads.
   - Constructing `AtTimeStamp` objects with the appropriate type and expression based on the provided input.

2. **Suggestions for Improvement:**
   - **Consolidation:** The added rules for handling timestamps and snapshots could be consolidated to reduce redundancy and improve maintainability. Consider refactoring the code to avoid repetition and enhance readability.
   - **Comments:** Adding comments to explain the purpose of the new rules and the significance of supporting stale reads can help other developers understand the changes more easily.
   - **Error Handling:** Ensure that appropriate error handling mechanisms are in place to handle unexpected inputs or edge cases related to stale read support.

3. **Security Concerns:**
   - **Input Sanitization:** It's crucial to validate and sanitize user inputs, especially when introducing new features like stale read support. Ensure that the parser handles inputs securely to prevent potential injection attacks or unexpected behavior.

4. **Code Quality:**
   - **Naming Conventions:** Maintain consistent naming conventions for variables, types, and functions to enhance code clarity and maintainability.
   - **Testing:** Verify that the changes are adequately tested to cover different scenarios related to stale read functionality.

### Overall, the pull request introduces necessary changes to support stale read in the parser. To enhance the quality of the codebase, consider consolidating the rules, adding comments for clarity, ensuring input validation, and maintaining code quality standards. Addressing these suggestions will not only improve the current implementation but also facilitate future maintenance and development efforts.

**Here are review comments for file pkg/sql/parsers/tree/table_name.go:**
### Pull Request Review:

#### Title:
The title of the pull request is clear and concise, indicating that the purpose of the changes is to support stale read in the plan.

#### Body:
The body of the pull request provides relevant information about the type of PR, the specific issue it addresses, and the reason for the changes. It mentions that the PR supports stale read in the plan, which gives context to the modifications made.

#### Changes in `table_name.go`:
1. **Addition of `ATTimeStampType` Enum:**
   - **Problem:** The addition of the `ATTimeStampType` enum without a clear explanation or context in the pull request description can make it difficult for reviewers to understand the purpose of this new type.
   - **Suggestion:** It would be beneficial to provide a brief explanation in the pull request description about why this enum was added and how it relates to supporting stale read in the plan.

2. **Modification in `AtTimeStamp` struct:**
   - **Problem:** The `Type` field was added to the `AtTimeStamp` struct without any usage or explanation in the `Format` method.
   - **Suggestion:** Ensure that the `Type` field is utilized appropriately in the `Format` method or provide a comment explaining its purpose if it's intended for future use.

3. **Addition of `String` method for `ATTimeStampType` enum:**
   - **Problem:** The `String` method for the `ATTimeStampType` enum was added without a clear indication of where it will be used.
   - **Suggestion:** If the `String` method is intended for logging or debugging purposes, consider adding comments to clarify its usage or provide an example of how it will be utilized.

#### Security Concerns:
- **No apparent security issues detected in the provided changes.**

#### Suggestions for Optimization:
- Ensure that the new enum and struct additions are utilized effectively in the existing codebase to support the intended functionality.
- Consider adding comments or documentation to explain the purpose and usage of the new additions for better code readability and maintainability.

#### Overall:
The changes made in the pull request seem to introduce new types and methods related to supporting stale read in the plan. To enhance the clarity and maintainability of the code, it would be beneficial to provide more context in the pull request description and ensure that the new additions are utilized appropriately. Additionally, consider adding comments to explain the purpose of the new enum and methods for better understanding by other developers.

**Here are review comments for file pkg/sql/plan/apply_indices.go:**
### Pull Request Review:

#### Title and Body:
The title of the pull request is clear and concise, indicating the purpose of the changes made. The body provides relevant information about the type of PR and the specific issue it addresses. It would be beneficial to include more details about the implementation approach or any potential impact of the changes in the body.

#### Changes in `apply_indices.go`:
1. **Addition of `ScanTS` Variable**:
   - **Problem**: The addition of the `ScanTS` variable in multiple places without proper context or explanation can lead to confusion. It's unclear why this variable is being introduced and how it affects the existing functionality.
   - **Suggestion**: Add comments or documentation explaining the purpose of the `ScanTS` variable and how it integrates with the existing code logic. Providing clear reasoning for its inclusion can help other developers understand its significance.

2. **Code Duplication**:
   - **Problem**: There is code duplication in the `applyIndicesForFiltersRegularIndex` and `applyIndicesForJoins` functions where the `ScanTS` assignment is repeated.
   - **Suggestion**: Consider refactoring the code to avoid duplication by extracting the common logic related to `ScanTS` assignment into a separate function or method. This can improve code readability and maintainability.

3. **Consistency in Variable Naming**:
   - **Problem**: Inconsistent variable naming conventions are observed, such as using `ts` in some places and `ScanTS` in others. Consistency in naming conventions enhances code clarity and maintainability.
   - **Suggestion**: Ensure consistent naming conventions are followed throughout the codebase. Choose a standard naming convention and apply it uniformly to variables for better code readability.

4. **Missing Error Handling**:
   - **Problem**: The code snippet provided does not show any error handling mechanisms for potential failures during the assignment or usage of the `ScanTS` variable.
   - **Suggestion**: Implement appropriate error handling mechanisms, such as checking for nil values or error returns when using the `ScanTS` variable to prevent unexpected behavior or crashes.

#### Security Concerns:
Since the provided code snippet does not reveal the complete context or the entire codebase, it is essential to ensure that the introduction of the `ScanTS` variable does not pose any security risks. It is crucial to validate the source of the timestamp and ensure that it is properly sanitized to prevent any injection attacks or unauthorized access.

### Overall Suggestions:
- Provide detailed comments or documentation explaining the purpose and usage of the `ScanTS` variable.
- Refactor the code to eliminate duplication and improve code maintainability.
- Ensure consistent variable naming conventions are followed for better code readability.
- Implement error handling mechanisms to handle potential failures related to the `ScanTS` variable.
- Conduct thorough testing to validate the functionality and security implications of the changes.

By addressing these points, the pull request can be enhanced in terms of clarity, maintainability, and security.

**Here are review comments for file pkg/sql/plan/mock.go:**
### Pull Request Review:

#### Title and Body:
The title of the pull request is clear and concise, indicating that the purpose of the change is to "support stale read in plan." The body of the pull request provides additional context by linking it to a specific issue (#14784) and stating the need for supporting stale read in the plan.

#### Changes in pkg/sql/plan/mock.go:
1. **Addition of ResolveSnapshotTsWithSnapShotName Function:**
   - The addition of the `ResolveSnapshotTsWithSnapShotName` function in the `MockCompilerContext` struct seems to be related to supporting stale reads in the plan.
   - **Issue:** The implementation of the function currently returns an empty string and nil error, which may not be the intended behavior for resolving snapshot timestamps.
   - **Suggestion:** Update the implementation of `ResolveSnapshotTsWithSnapShotName` to properly handle resolving snapshot timestamps based on the provided `snapshotName`.

2. **Optimization Opportunity:**
   - Since the function `ResolveSnapshotTsWithSnapShotName` is added to support stale reads, it would be beneficial to ensure that the function is implemented correctly to provide the intended functionality.
   - **Suggestion:** Consider adding logic to `ResolveSnapshotTsWithSnapShotName` that retrieves the appropriate snapshot timestamp based on the provided `snapshotName` parameter.

3. **Code Quality:**
   - It's important to ensure that the new function `ResolveSnapshotTsWithSnapShotName` is well-documented with clear comments explaining its purpose, parameters, and expected behavior.
   - **Suggestion:** Add comments to the `ResolveSnapshotTsWithSnapShotName` function to improve code readability and maintainability.

#### Security Concerns:
- As the current implementation of `ResolveSnapshotTsWithSnapShotName` returns an empty string without any validation or logic, there is a potential security risk if this function is intended to handle sensitive data or operations related to stale reads.
- **Security Suggestion:** Ensure that proper validation and logic are implemented within `ResolveSnapshotTsWithSnapShotName` to handle snapshot timestamp resolution securely and accurately.

### Overall Recommendations:
- Verify and update the implementation of `ResolveSnapshotTsWithSnapShotName` to correctly handle snapshot timestamp resolution.
- Add comprehensive comments to the new function for better code understanding.
- Address any security concerns related to handling sensitive data within the function.

By addressing these points, the pull request can enhance the support for stale reads in the plan while maintaining code quality and security standards.

**Here are review comments for file pkg/sql/plan/query_builder.go:**
### Pull Request Review:

#### Title and Body:
The title of the pull request is clear and concise, indicating that the PR is about supporting stale read in the plan. The body provides information about the type of PR and the issue it fixes, which is good practice.

#### Changes in `pkg/sql/plan/query_builder.go`:
1. **Addition of `timestamp` Package Import:**
   - The addition of `timestamp` package import is necessary for the new functionality related to timestamp handling.

2. **Addition of `timestamp.Timestamp` Variable:**
   - A new variable `ts` of type `timestamp.Timestamp` is added to handle timestamp values.

3. **Changes in `buildTable` Function:**
   - In the `buildTable` function, a new variable `ts` is declared to store timestamp values.
   - If `tbl.AtTsExpr` is not nil, it calls `resolveTsHint` function to resolve the timestamp hint.
   - The resolved timestamp is then assigned to the `ScanTS` field in the `TableReader` struct.

4. **Addition of `resolveTsHint` Function:**
   - A new function `resolveTsHint` is added to handle the resolution of timestamp hints.
   - It parses the timestamp hint expression and extracts the timestamp value based on the type of hint.
   - It handles different cases like `ATTIMESTAMPTIME` and `ATTIMESTAMPSNAPSHOT` to parse timestamps accordingly.

#### Suggestions for Improvement:
1. **Error Handling:**
   - Ensure robust error handling throughout the code, especially in the `resolveTsHint` function where various parsing and resolution operations are performed. Make sure to handle and propagate errors appropriately.

2. **Code Optimization:**
   - Consider refactoring the `resolveTsHint` function for better readability and maintainability. Break down complex logic into smaller, more manageable functions if necessary.

3. **Consistency:**
   - Ensure consistency in variable naming and coding style throughout the file to maintain code readability and coherence.

4. **Documentation:**
   - Add comments and documentation to explain the purpose and functionality of the new code additions, especially in the `resolveTsHint` function to aid future developers in understanding the logic.

5. **Security Considerations:**
   - Ensure that the handling of timestamp values is secure, especially when parsing and resolving user-provided input. Validate input to prevent any potential security vulnerabilities like injection attacks.

6. **Testing:**
   - Write comprehensive unit tests to cover the new functionality added in this PR, especially for the `resolveTsHint` function to ensure its correctness and reliability.

By addressing the above points, the codebase can be improved in terms of readability, maintainability, and security. Additionally, thorough testing will help ensure the robustness of the new features introduced in this pull request.

**Here are review comments for file pkg/sql/plan/types.go:**
### Pull Request Review:

#### Title and Body:
The title of the pull request is clear and concise, indicating that the purpose of the change is to "support stale read in plan." The body of the pull request provides additional context by referencing the related issue #14784. However, it lacks detailed information on why supporting stale read in the plan is necessary or how it addresses the mentioned issue. It would be beneficial to include more specific details in the body to help reviewers understand the rationale behind the change.

#### Changes in `pkg/sql/plan/types.go`:
1. **Addition of `ResolveSnapshotTsWithSnapShotName` method:**
   - The addition of the `ResolveSnapshotTsWithSnapShotName` method in the `CompilerContext` interface is a significant change. However, the method name contains a typo (`SnapShot` instead of `Snapshot`).
   - **Issue:** The method name inconsistency may lead to confusion and should be corrected for consistency and readability.
   - **Suggestion:** Correct the method name to `ResolveSnapshotTSWithSnapshotName` for consistency and clarity.

2. **Code Comment:**
   - The existing comment for the `ReplacePlan` method is incomplete and lacks details on the expected return value.
   - **Suggestion:** Consider updating the comment for the `ReplacePlan` method to provide a clear description of the return value and any other relevant information.

3. **Code Formatting:**
   - The added method lacks proper formatting with respect to surrounding code (spacing, indentation).
   - **Suggestion:** Ensure consistent code formatting by aligning the new method declaration with the existing code style for better readability and maintainability.

### Overall Suggestions:
- Provide a more detailed explanation in the pull request body to clarify the purpose and impact of the changes.
- Correct the method name typo in `ResolveSnapshotTsWithSnapShotName` for consistency.
- Enhance the code comment for the `ReplacePlan` method to improve clarity.
- Maintain consistent code formatting for better code readability.

By addressing these suggestions, the pull request can be improved in terms of clarity, consistency, and maintainability.

**Here are review comments for file pkg/sql/plan/types_mock.go:**
### Pull Request Review:

#### Title and Body:
The title of the pull request is clear and concise, indicating that the purpose of the changes is to support stale read in the plan. The body of the pull request provides additional context by mentioning the related issue number (#14784) and the need for supporting stale read in the plan. It would be beneficial to include more details about the changes made and the impact they will have on the codebase in the body of the pull request.

#### Changes in `pkg/sql/plan/types_mock.go`:
1. **Addition of `ResolveSnapshotTsWithSnapShotName` method:**
   - The addition of this method seems to be related to supporting stale read in the plan, as mentioned in the pull request.
   - It is important to ensure that the implementation of this method correctly handles snapshot names and any potential errors that may arise.
   - Consider adding comments or documentation to explain the purpose of this method and how it fits into the overall functionality of the codebase.

2. **MockRecorder for `ResolveSnapshotTsWithSnapShotName` method:**
   - The addition of the `ResolveSnapshotTsWithSnapShotName` method in the MockRecorder indicates that this method is expected to be called during testing.
   - Verify that the MockRecorder is correctly set up to record and track calls to this method for testing purposes.
   - Ensure that the MockRecorder accurately reflects the behavior of the `ResolveSnapshotTsWithSnapShotName` method.

#### Suggestions for Improvement:
1. **Documentation and Comments:**
   - Add comments to explain the purpose of the new `ResolveSnapshotTsWithSnapShotName` method and its parameters.
   - Consider adding documentation to describe how stale read is being supported in the plan and how this new method contributes to that functionality.

2. **Error Handling:**
   - Ensure that proper error handling is implemented in the `ResolveSnapshotTsWithSnapShotName` method to handle any potential errors that may occur during its execution.
   - Consider adding error checks and returning meaningful error messages to provide better feedback in case of failures.

3. **Testing:**
   - Write unit tests to validate the behavior of the new `ResolveSnapshotTsWithSnapShotName` method and ensure that it functions as expected.
   - Verify that the MockRecorder setup is correct and that it accurately captures calls to the method for testing purposes.

4. **Optimization:**
   - Review the implementation of the new method to ensure it follows best practices and is optimized for performance.
   - Consider refactoring or simplifying the code if possible to improve readability and maintainability.

By addressing the suggestions mentioned above and ensuring thorough testing and documentation, the pull request can be enhanced to improve the quality and reliability of the codebase.

**Here are review comments for file pkg/vm/engine/memoryengine/compiler_context.go:**
### Pull Request Review:

#### Title:
The title of the pull request is clear and concise, indicating that the purpose of the changes is to "support stale read in plan."

#### Body:
The body of the pull request provides some context by mentioning the type of PR (Feature), the specific issue it fixes (issue #14784), and the reason for the changes ("support stale read in plan"). It would be beneficial to include more details about the implementation approach or any potential impact of the changes.

#### Changes in `compiler_context.go`:
1. **New Function `ResolveSnapshotTsWithSnapShotName`**:
   - The addition of the `ResolveSnapshotTsWithSnapShotName` function is a good step towards supporting stale reads.
   - The function is currently not implemented and panics with "implement me." This is not a recommended practice for production code as it can lead to runtime errors.
   - **Suggestions**:
     - Implement the logic inside `ResolveSnapshotTsWithSnapShotName` function to handle resolving snapshot timestamps based on the snapshot name provided.
     - Instead of panicking, consider returning an error or a default value if the implementation is not yet complete.

2. **Unimplemented Function `SetQueryingSubscription`**:
   - The existing `SetQueryingSubscription` function still has a placeholder comment and panics with "implement me."
   - **Suggestions**:
     - Implement the logic for `SetQueryingSubscription` function to set querying subscription based on the provided `SubscriptionMeta`.
     - Avoid using panic for unimplemented functions; consider returning an error or logging a message instead.

#### Overall Suggestions:
- Implement the logic for the new `ResolveSnapshotTsWithSnapShotName` function to handle snapshot timestamp resolution.
- Complete the implementation of the existing `SetQueryingSubscription` function to set querying subscriptions.
- Replace panics with appropriate error handling mechanisms to prevent unexpected crashes in production code.
- Consider adding comments/documentation to explain the purpose and usage of the new and existing functions for better code readability.

By addressing the suggestions provided and completing the implementation of the functions, the pull request can enhance the support for stale reads in the plan while ensuring code reliability and maintainability.

**Here are review comments for file proto/plan.proto:**
### Pull Request Review:

#### Title and Body:
The title of the pull request is clear and concise, indicating that the purpose of the changes is to "support stale read in plan." The body of the pull request provides additional context by mentioning the related issue (#14784) and the need for supporting stale read in the plan.

#### Changes in `proto/plan.proto`:
1. **Addition of `timestamp.proto` import:**
   - The pull request adds an import statement for `timestamp.proto` in the `plan.proto` file.
   - **Issue:** While adding the import is necessary for using the `timestamp.Timestamp` type, it is essential to ensure that the imported file is secure and does not introduce any vulnerabilities.
   - **Suggestion:** Before merging the changes, conduct a security review of the `timestamp.proto` file to ensure it does not contain any malicious code or security vulnerabilities.

2. **Addition of `scanTS` field in `Node` message:**
   - A new field `scanTS` of type `timestamp.Timestamp` is added to the `Node` message.
   - **Issue:** It is crucial to validate the usage of the `scanTS` field to prevent any potential misuse or incorrect handling of timestamps.
   - **Suggestion:** Add comments or documentation explaining the purpose and usage of the `scanTS` field to guide developers on how to correctly utilize this new feature.

#### Overall Suggestions:
1. **Security Review:**
   - Perform a thorough security review of the newly imported `timestamp.proto` file to ensure it does not introduce any security vulnerabilities into the codebase.

2. **Documentation:**
   - Provide clear documentation or comments explaining the rationale behind supporting stale read in the plan and how the new `scanTS` field should be used.

3. **Testing:**
   - Write test cases to validate the functionality of the changes related to supporting stale read in the plan and the usage of the `scanTS` field.

4. **Code Review:**
   - Have the changes reviewed by team members to ensure adherence to coding standards and best practices.

By addressing the security concerns, documenting the changes, writing tests, and conducting a thorough code review, the pull request can be optimized for merging into the codebase effectively.

**Review by reusee — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]

**Review by daviszhen — APPROVED** [review]

**Review by iamlinjunhong — APPROVED** [review]

**Review by ouyuanning — APPROVED** [review]

**Review by m-schen — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Review by heni02 — APPROVED** [review]

**Pull request merged**
